Replace Connman and BanMan globals with NodeContext local

ryanofsky opened the pull request:
This change is mainly a naming / organization change intended to simplify #10102. It:

- Renames struct InitInterfaces to struct NodeContext and moves it from
  src/init.h to src/node/context.h. This is a cosmetic change intended to make
  the point of the struct more obvious.

- Gets rid of BanMan and ConnMan globals making them NodeContext members
  instead. Getting rid of these globals has been talked about in past as a way
  to implement testing and simulations. Making them NodeContext members is a
  way of keeping them accessible without the globals.

- Splits g_rpc_interfaces global into g_rpc_node and g_rpc_chain globals. This
  better separates node and wallet rpc methods. Node RPC methods should have
  access NodeContext, while wallet RPC methods should only have indirect access
  to node functionality via interfaces::Chain.

- Adds NodeContext& references to interfaces::Chain class and the
  interfaces::MakeChain() function. This is needed to access ConnMan and BanMan
  instances without the globals.

- Gets rid of redundant Node and Chain instances in Qt tests. This is
  needed due to the previous MakeChain change, and also makes test setup a
  little more straightforward. More cleanup could be done in the future, but it
  will require deduplication of bitcoind, bitcoin-qt, and TestingSetup init
  code.

ryanofsky commented:
I'm looking for concept ACKS before detailed review at the moment. Since this a naming and refactoring change, it'd be great if people could take few minutes and skim the diff, and say whether the new code looks ugly or ok, and offer any suggestions. I'm happy to take suggestions and also split this up more.

promag commented on src/interfaces/chain.cpp:
```diff
@@ -375,9 +377,16 @@ class ChainImpl : public Chain
+    {
+        ChainClients& clients = m_node.chain_clients;
+        ChainClients::iterator it = clients.emplace(clients.end(), client);
+        return MakeHandler([&clients, it] { clients.erase(it); });
```
Is it better to capture member `chain_clients` than `this`?

Also, it would be nice to explain that a `std::list` is used otherwise this `.erase` would fail.

promag reviewed (commented):
Concept ACK.

practicalswift commented:
Concept ACK

MarcoFalke commented:
> Getting rid of these globals has been talked about in past as a way
to implement testing and simulations.

Fine with me, but this seems only like a first step, since other globals such as `::mempool` are still global. and not members of the Node class.

ryanofsky reviewed (commented):
Thanks for feedback!

Rebased 124bbcabe314c2e4a48ad61a69c03ee55e633e37 -> 7225ebebdfeac3d365c1490f1b737403488005d7 ([pr/noglob.3](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.3) -> [pr/noglob.4](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.4)) due to conflict with #16787. Also added requested comment and no-wallet build fix.

re: https://github.com/bitcoin/bitcoin/pull/16839#issuecomment-529941376
> Fine with me, but this seems only like a first step, since other globals such as ::mempool are still global. and not members of the Node class.

Yes, it's not all or nothing. It's perfectly fine for ::mempool to remain a global, but if we want to write multiple-instance tests for code currently using ::mempool, we would have the option of moving it to the Node struct and still accessing it conveniently.

jamesob commented:
Concept ACK. This looks like a step in the right direction in terms of cutting down on globals and moving towards being able to test multi-node behavior without having to rely on the functional tests.

Is your intent to split this into separate commits at some point for detailed review?

ryanofsky commented:
> Is your intent to split this into separate commits at some point for detailed review?

Could go either way. These are basically boring text replacements that should be about as much effort to review split up or joined, but I'm happy to split it up. I think I could split it up into 4 or 5 commits roughly matching the bullet points in the description. Overall diff would be a little bigger because some lines (`chain_clients` lines) would change more than once.

jamesob commented:
> Could go either way.

I'm fine reviewing either, though I guess if it were easy to do some scripted-diffs that would be nice.

ariard commented on src/interfaces/node.h:
```diff
@@ -254,6 +255,10 @@ class Node
     virtual std::unique_ptr<Handler> handleNotifyHeaderTip(NotifyHeaderTipFn fn) = 0;
+
+    //! Return pointer to internal chain interface, useful for testing.
+    //! Only works if node interface is being called in same process as implementation.
```
I find comment fuzzy, do you mean only works if node interface is set by the process of the implementation ?

ariard commented on src/node/node.h:
```diff
@@ -0,0 +1,41 @@
+//! Node struct containing chain state and connection state.
+//!
+//! More state could be moved into this struct to eliminate global variables,
+//! and allow creating multiple instances of validation and networking objects
```
So you would create multiple Node in one test process, connect them and let them talk maybe with random introduction of interferences in-between ?

ariard commented on src/node/transaction.cpp:
```diff
@@ -14,12 +14,12 @@
 #include <future>
 
-TransactionError BroadcastTransaction(const CTransactionRef tx, std::string& err_string, const CAmount& max_tx_fee, bool relay, bool wait_callback)
+TransactionError BroadcastTransaction(const CTransactionRef tx, std::string& err_string, const CAmount& max_tx_fee, bool relay, bool wait_callback, CConnman* connman)
```
Couldn't BroadcastTransaction be directly a method of Node instead of passing CConnman ? Maybe sendrawtransaction should be given g_rpc_chain instead of g_rpc_node, overall it's seems to use the chain as a server too

MarcoFalke replied:
> Yes, but I don't think that would be a good idea. I think interface classes should be a thin layer on top of wallet and node code and not a place where real functionality is implemented. The model is:

I don't understand this answer. `src/node/transaction.cpp` is "node code" as I understand it and not merely some interface. All the other global state accessed in this function (chainstate, mempool) should eventually be a node member. So the suggestion by @ariard to make this a node member function makes sense to me.

ariard replied:
Thanks got your point. Was doing far out-of-scope thinking, `sendrawtransaction` to your local node isn't great for privacy, it could be a binary to connect to any node to broadcast instead of a RPC call. But yes, in that case it wouldn't even use any interface..

ariard commented on src/rpc/blockchain.h:
```diff
@@ -46,4 +47,9 @@ UniValue blockheaderToJSON(const CBlockIndex* tip, const CBlockIndex* blockindex
 void CalculatePercentilesByWeight(CAmount result[NUM_GETBLOCKSTATS_PERCENTILES], std::vector<std::pair<CAmount, int64_t>>& scores, int64_t total_weight);
 
+//! Pointer to node state that needs to be declared as a global to be accessible
+//! RPC methods. Due to limitations of the RPC framework, there's currently no
```
How to override RPC framework limitations ? Encapsulating RPC functions in its own structure like it has been done for the P2P stack with CConnman ?

ariard commented on src/rpc/mining.cpp:
```diff
@@ -424,10 +425,10 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid mode");
 
-    if(!g_connman)
+    if(!g_rpc_node->connman)
```
IMO not sure why getblocktemplate need to access state of P2P stack ? I would guess the functionality would be useless without a mempool to populate the template and being connected increase the likeliness of a fresh mempool but I could also provide mempool.dat to my node...
If you remove it, you can also move g_rpc_node declaration in src/rpc/net.h.

ariard reviewed (commented):
Great concept ACK! See comments on minor points, also I think maybe the Handler/Cleanup handler logic could be simplified in future work, didn't spend time on the how yet..

ryanofsky committed:
scripted-diff: Rename InitInterfaces to NodeContext

-BEGIN VERIFY SCRIPT-
s() { git grep -l "$1" src | xargs sed -i "s/$1/$2/g"; }

s 'struct InitInterfaces'              'struct NodeContext'
s 'InitInterfaces interfaces'          'NodeContext node'
s 'InitInterfaces& interfaces'         'NodeContext\& node'
s 'InitInterfaces m_interfaces'        'NodeContext m_context'
s 'InitInterfaces\* g_rpc_interfaces'  'NodeContext* g_rpc_node'
s 'g_rpc_interfaces = &interfaces'     'g_rpc_node = \&node'
s 'g_rpc_interfaces'                   'g_rpc_node'
s 'm_interfaces'                       'm_context'
s 'interfaces\.chain'                  'node.chain'
s '\(AppInitMain\|Shutdown\|Construct\)(interfaces)' '\1(node)'
s 'init interfaces' 'chain clients'
-END VERIFY SCRIPT-

ryanofsky committed:
MOVEONLY: Move NodeContext struct to node/context.h

ryanofsky committed:
Pass NodeContext, ConnMan, BanMan references more places

So g_connman and g_banman globals can be removed next commit.

ryanofsky committed:
scripted-diff: Remove g_connman, g_banman globals

-BEGIN VERIFY SCRIPT-
sed -i 's:#include <interfaces/chain.h>:#include <banman.h>\n#include <interfaces/chain.h>\n#include <net.h>\n#include <net_processing.h>:' src/node/context.cpp
sed -i 's/namespace interfaces {/class BanMan;\nclass CConnman;\nclass PeerLogicValidation;\n&/' src/node/context.h
sed -i 's/std::unique_ptr<interfaces::Chain> chain/std::unique_ptr<CConnman> connman;\n    std::unique_ptr<PeerLogicValidation> peer_logic;\n    std::unique_ptr<BanMan> banman;\n    &/' src/node/context.h
sed -i '/std::unique_ptr<[^>]\+> \(g_connman\|g_banman\|peerLogic\);/d' src/banman.h src/net.h src/init.cpp
sed -i 's/g_connman/m_context.connman/g' src/interfaces/node.cpp
sed -i 's/g_banman/m_context.banman/g' src/interfaces/node.cpp
sed -i 's/g_connman/m_node.connman/g' src/interfaces/chain.cpp src/test/setup_common.cpp
sed -i 's/g_banman/m_node.banman/g' src/test/setup_common.cpp
sed -i 's/g_connman/node.connman/g' src/init.cpp src/node/transaction.cpp
sed -i 's/g_banman/node.banman/g' src/init.cpp
sed -i 's/peerLogic/node.peer_logic/g' src/init.cpp
sed -i 's/g_connman/g_rpc_node->connman/g' src/rpc/mining.cpp src/rpc/net.cpp src/rpc/rawtransaction.cpp
sed -i 's/g_banman/g_rpc_node->banman/g' src/rpc/net.cpp
sed -i 's/std::shared_ptr<CWallet> wallet =/node.context()->connman = std::move(test.m_node.connman);\n    &/' src/qt/test/wallettests.cpp
-END VERIFY SCRIPT-

ryanofsky committed:
Avoid using g_rpc_node global in wallet code

Wallet code should use interfaces::Chain and not directly access to node state.

Add a g_rpc_chain replacement global for wallet code to use, and move
g_rpc_node definition to a libbitcoin_server source file so there are link
errors if wallet code tries to access it.

ryanofsky reviewed (commented):
> I'm fine reviewing either, though I guess if it were easy to do some scripted-diffs that would be nice.

Thanks, I split this up into smaller commits now (two of them scripted diffs)

Updated a3dc53304a4ccd0996e51322bee40572b5af7d25 -> 7b65a5356ea107af345b980a53e7e2d669711461 ([pr/noglob.5](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.5) -> [pr/noglob.6](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/noglob.5..pr/noglob.6))

ryanofsky commented:
Could remove the "Needs conceptual review" tag, I think, since this has had a few concept ACKS.

I updated this to drop the last commit 7b65a5356ea107af345b980a53e7e2d669711461, which turned out not to be very related to the other changes, and just complicated the PR

Updated 7b65a5356ea107af345b980a53e7e2d669711461 -> 437a5e439fff50567503c0d76baaba50ebbfbd8f ([pr/noglob.6](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.6) -> [pr/noglob.7](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/noglob.6..pr/noglob.7)) dropping the last commit
Rebased 437a5e439fff50567503c0d76baaba50ebbfbd8f -> 396544eb9e7bfbcf41ef7f541b10ffd004603b5b ([pr/noglob.7](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.7) -> [pr/noglob.8](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.8)) due to minor #include line conflict with #16512

MarcoFalke commented on src/node/node.h:
```diff
@@ -20,6 +23,9 @@ class ChainClient;
 //! and linking outside of bitcoind for simulation or testing.
 struct Node
 {
+    std::unique_ptr<CConnman> connman;
```
I think we should be using `struct` only for the dumbest data structures and not complex classes (such as the node). So this should be changed to `m_connman`

MarcoFalke replied:
Thanks for clarifying. Could add that sentence to the doxygen comment of `Node` for people like me?

maflcko reviewed (commented):
ACK (looked at the GitHub diff)

MarcoFalke commented on src/node/transaction.cpp:
```diff
@@ -14,12 +14,12 @@
 #include <future>
 
-TransactionError BroadcastTransaction(const CTransactionRef tx, std::string& err_string, const CAmount& max_tx_fee, bool relay, bool wait_callback)
+TransactionError BroadcastTransaction(const CTransactionRef tx, std::string& err_string, const CAmount& max_tx_fee, bool relay, bool wait_callback, CConnman* connman)
```
I guess I'd prefer to pass in the node here, so that in the future when mempool and chainstate are members of the node struct, this doesn't need to change again.

```suggestion
TransactionError BroadcastTransaction(const CTransactionRef tx, std::string& err_string, const CAmount& max_tx_fee, bool relay, bool wait_callback, Node& node)
```

ryanofsky reviewed (commented):
Thanks for the review!

Updated 396544eb9e7bfbcf41ef7f541b10ffd004603b5b -> dab8d3e2c9cd9bfc92a4520edf6bf3374a9fe4f5 ([pr/noglob.8](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.8) -> [pr/noglob.9](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/noglob.8..pr/noglob.9)) with suggested changes: better Node comment, BroadcastTransaction argument change.

ryanofsky commented:
FYI maintainers, "Needs conceptual review" label was removed but this is still showing up in the "Chasing Concept ACK" column from https://github.com/bitcoin/bitcoin/projects/8.

ryanofsky commented:
Rebased dab8d3e2c9cd9bfc92a4520edf6bf3374a9fe4f5 -> ae5d88f8e08def49c7afdda040d6bb9fe73a0e3d ([pr/noglob.9](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.9) -> [pr/noglob.10](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.10)) due to conflict with #16912

jamesob commented:
ACK https://github.com/bitcoin/bitcoin/pull/16839/commits/ae5d88f8e08def49c7afdda040d6bb9fe73a0e3d

Read through the code partially locally (to verify move-onlys) and on Github. Seems pretty uncontroversial and nicely split into scripted-diffs. I can do some manual testing tomorrow if need be. This `Node` object has a lot of promise for testing instead of having to push around globals.

ryanofsky commented:
Rebased ae5d88f8e08def49c7afdda040d6bb9fe73a0e3d -> 50e0bb2df33cc47be10cf8e26ea7d9ce1d2e9704 ([pr/noglob.10](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.10) -> [pr/noglob.11](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.11)) due to minor conflict (in qt test comment) with #17015

ariard commented on src/interfaces/node.cpp:
```diff
@@ -259,12 +259,12 @@ class NodeImpl : public Node
     {
         std::shared_ptr<CWallet> wallet;
-        WalletCreationStatus status = CreateWallet(*m_interfaces.chain, passphrase, wallet_creation_flags, name, error, warning, wallet);
+        WalletCreationStatus status = CreateWallet(*m_node.chain, passphrase, wallet_creation_flags, name, error, warning, wallet);
```
Do `Node` really need a reference to `Chain`, `ChainImpl` has already a reference to `Node` that's a bit confusing ? I mean `createWallet` and `loadWallet` couldn't be method of `Wallet` interface instead ? Would it work for the GUI ?

More generally, if we refactor RPC code to encapsulate in a `RPCServer` class, I see a `WalletRPC` subclass with a `Chain` member and a `NodeRPC` subclass with a `Node` member. And `Node` maybe initialize differently to avoid access to uneeded state which would address some concerns of https://github.com/bitcoin/bitcoin/pull/16839#discussion_r325866327

ariard replied:
Thanks for answers (sorry for delay), wasn't implying to do changes in this PR. Apart of RPC framework which need design goals, I think other one would make code nicer in future PRs, bookmarked.

ariard commented on src/rpc/blockchain.h:
```diff
@@ -46,4 +47,9 @@ UniValue blockheaderToJSON(const CBlockIndex* tip, const CBlockIndex* blockindex
+//! Pointer to node state that needs to be declared as a global to be accessible
+//! RPC methods. Due to limitations of the RPC framework, there's currently no
+//! direct way to pass in state to RPC methods without globals.
+extern Node* g_rpc_node;
```
nit: Wouldn't shock me to move it to a new src/rpc/node.h libbitcoin_server source file but maybe too early

ariard commented on src/node/node.h:
```diff
@@ -13,11 +13,25 @@ class Chain;
+//! and make code more modular and testable. The struct isn't intended to have
+//! any member functions. It should just be a collection of references that can
+//! be used without pulling in unwanted dependencies or functionality.
 struct Node
```
nit: maybe it should be name NodeState to avoid confusion with `Node` the already-there interface class?

ariard replied:
>  I guess I have some aversion to adding suffixes like Info Data State to type names because you could add those suffixes to pretty much any type name just making it more verbose without changing the meaning.

Ah, you're right but most of the times it's the cheapest way to avoid name collisions!

+1 for `NodeInterface` if we have opportunity later

ariard reviewed (approved):
ACK 50e0bb2, run tests and thoroughly read code.

I find the interwining of `Node` and `Chain` a bit confusing but good enough for now.

ryanofsky reviewed (commented):
Thanks for the review!

Rebased 50e0bb2df33cc47be10cf8e26ea7d9ce1d2e9704 -> 9c4a2526615e6db54b49cf84ac77e165c3c9cbf3 ([pr/noglob.11](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.11) -> [pr/noglob.12](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.12)) due to conflict with #15437

re: https://github.com/bitcoin/bitcoin/pull/16839#pullrequestreview-298995605

> I find the interwining of `Node` and `Chain` a bit confusing but good enough for now.

In your comments you refer to `Chain` but in the code it is always `interfaces::Chain`. I wonder if renaming `interfaces::Chain` to `ChainInterface` would make it clearer that this class is an external interface used outside the node (by the wallet) to access chain state and not something that has to do with implementing node functionality or storing node data.

MarcoFalke commented:
Looks ready for merge after rebase

ryanofsky commented:
Rebased 34a7a50dbeec1ee36b18fa0d1a0ffbf7a1907f98 -> 362ded410b8cb1104b7ef31ff8488fec4824a7d5 ([pr/noglob.15](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.15) -> [pr/noglob.16](https://github.com/ryanofsky/bitcoin/commits/pr/noglob.16)) due to conflict with #16202

Also renamed `struct Node` to `struct NodeContext`, since it should also be useful to have a `WalletContext` later to put things like `cs_wallets`, `vpwallets`, and the `interfaces::Chain` pointer shared between multiple wallets.

laanwj commented:
ACK 362ded410b8cb1104b7ef31ff8488fec4824a7d5

ariard commented:
Post-humous ACK, diff from 50e0bb2 is renaming `struct Node`, integration of bip61 removal and some comments changed.

> I wonder if renaming interfaces::Chain to ChainInterface would make it clearer that this class is an external interface used outside the node (by the wallet) to access chain state and not something that has to do with implementing node functionality or storing node data.

Agree may have avoid us confusion [there](https://github.com/bitcoin/bitcoin/pull/16839#discussion_r324425941).
